### DISCUSSION TITLE
How to use level in LAB colorspace ?

### DISCUSSION THREAD

**Discussion by harlock974** [ImageMagick/ImageMagick#3729] [Help] [answered]
Hello
I wish to adjust levels in either A or B channel in LAB colorspace.
I thought at first to use something like this : 

`convert test.png -colorspace LAB -channel A -level 20%,100% result.png`

But it doesn't work, A is here Alpha channel and not A from LAB.
Is there a way to do it ?

**Comment by snibgo** [comment]
What IM version, on what platform?

"convert -list channel" lists the channel names you can use. For Lab, I would use "0", "1" and "2".

**Reply by harlock974** [reply]
IM version 6.9.7-4 on Linux Mint.
With "-list channel" there is the usual names (Red, Green, etc...) and numbers from 0 to 5.

If channel 2 is LAB B I should have a blue / yellow change but : 

-  `convert test.png -colorspace LAB -channel 2 -level 40%,100% result.png` shift to cyan
- `convert test.png -colorspace LAB -channel 2 -level 0%,60% result.png` shift to magenta
- `convert test.png -colorspace LAB -channel 1 -level 0%,60% result.png` brightens the image
- `convert test.png -colorspace LAB -channel 1 -level 40%,100% result.png` darkens the image

and no change if I use channel 0...

**Comment by snibgo** [comment]
I suggest that whenever you have "-channel X" you also have a corresponding "+channel" to enable all channels.

I suggest you explicitly convert back to sRGB, with "-colorspace sRGB". As PNG cannot record Lab, IM will do this automatically, but I think explicit conversion is better practise.

You don't show test.png so I can't comment on your results, but they work fine for me. Instead of using "-level, try "-evaluate set 100%" or "-evaluate set 0".

For example:
```
convert test.png -colorspace Lab -channel 2 -evaluate set 100% +channel -colorspace sRGB result.png
```

**Comment by harlock974** [comment, marked as answer]
Thanks Snibgo for the answer.
As an example, this basic image : 

![elsa](https://user-images.githubusercontent.com/6268242/119371572-a18fc600-bcc7-11eb-846c-44c4798685f0.jpg)

With Gimp I separate channels in the LAB colorspace, and apply a 75% curve to A channel : 

![curve](https://user-images.githubusercontent.com/6268242/119371920-eddb0600-bcc7-11eb-8473-dfe32d264243.png)

This increases magenta, as expected : 

![elsa2](https://user-images.githubusercontent.com/6268242/119372177-14993c80-bcc8-11eb-98a8-65c8f333b0b3.jpg)

I notice I could get a quite close, but not exactly the same, result with : 

`convert elsa.jpg -colorspace Lab -channel 2 -level 0%,75% +channel -colorspace sRGB essai.jpg`

![essai](https://user-images.githubusercontent.com/6268242/119372511-748fe300-bcc8-11eb-8132-3bdc21adadc7.jpg)

`+channel` and `-colorspace sRGB` seem optional as the result is the same without.

Channel 4 is close to Lab B and channel 1 to Lab Lightness. The difference with Gimp is small so I finally could process my images with Imagemagick. Thanks !

**Reply by fmw42** [reply]
There is no channel 4 for LAB in ImageMagick if no alpha channel. Channels are numbered from 0. So 0 is L, 1 is A and 2 is B. The alpha channel needs to be processed using -channel A, which is not the same as the A channel in LAB. I think the alpha channel would be channel 4.

In GIMP, your 75% change is non-linear. Imagemagick's -level 0x75% is linear. So they would not be exactly the same.

**Comment by snibgo** [comment]
> With Gimp I separate channels in the LAB colorspace, and apply a 75% curve to A channel :

The graph shows that you have transformed the A channel with a _curve_. IM's "-level" operation is linear, a straight line, not a curve.

Gimp has an option to use a linear transformation.

**Reply by harlock974** [reply]
Hi

> The graph shows that you have transformed the A channel with a curve. IM's "-level" operation is linear, a straight line, not a curve.

Yes I know, but the curve here is nearly linear. Anyway that could explain the slight differences.

> There is no channel 4 for LAB in ImageMagick if no alpha channel. Channels are numbered from 0. So 0 is L, 1 is A and 2 is B. The alpha channel needs to be processed using -channel A, which is not the same as the A channel in LAB. I think the alpha channel would be channel 4.

I can't explain this but I haven't any change in picture using channel 0 so I suppose it's alpha in this case. Using channel 2 and 4 on difficult underwater pictures gives similar result to Gimp processing with Lab A and B.

An example of color improvement : 
The original image is taken with a Gopro Hero 3+ at 16 meters depth : 

![before](https://user-images.githubusercontent.com/6268242/119441178-bc4d5380-bd36-11eb-8474-45342b40b4f2.jpg)

This processing is applied : 

`mogrify -auto-level -channel GB -sigmoidal-contrast 1,100% -format png image.jpg`
`mogrify -colorspace Lab -channel 2 -level -20%,100% -channel 4 -level 17%,90% image.png`

And the result : 

![after](https://user-images.githubusercontent.com/6268242/119441403-2534cb80-bd37-11eb-9105-c912c4e01717.jpg)

**Reply by fmw42** [reply]
Your input image is JPG, so it has no alpha channel, though you can set that to be opaque and change it. But I doubt doing that is equivalent to anything meaningful in regard to what you are doing in GIMP. Just playing around blindly with channels is not really a good idea even if you accidentally get something close on one image. It might not be meaningful on any other image.

**Reply by fmw42** [reply]
You also need to set +channel after processing one channel and before selecting some other channel or you will be process the first channel all over again.

**Comment by fmw42** [comment]
Perhaps this is what you want. LAB is biased. So you have to unbias, process, then bias again. Also you are moving your curve from 75 to 100. So that would be +level 0,75%.  Or you could use -level 0,125% to move 100 to 125.

![divers](https://user-images.githubusercontent.com/13458218/119445203-4a96ea80-bce1-11eb-9ef5-c8be4423d7b6.jpg)


```
mogrify -auto-level -channel GB -sigmoidal-contrast 1,100% +channel -format png divers.jpg
mogrify -colorspace Lab -channel 1,2 -bias -50% +level 0,75% -bias 50% +channel divers.png
```

![divers](https://user-images.githubusercontent.com/13458218/119445179-44087300-bce1-11eb-944f-2e9fa8e20fa3.png)

**Reply by harlock974** [reply]
Thanks fmw42 for the tips, it helps me to understand what happened.

Channel number mystery is solved. I tried another version of IM and, for LAB colorspace : 

* IM version 6.9.7-4     : L = 1, A = 2, B = 4
* IM version 6.9.10-23 : L = 0, A = 1, B = 2

About the use of `+channel` : according to the manual (and my own experiments), `-channel` doesn't seem cumulative, so in : 

`-channel 1 -action1 -channel2 -action2 +channel -action3`

* action1 is applied to channel 1 only
* action2 is applied to channel 2 only
* action3 is applied to default channels (RGBK,sync)

**Reply by fmw42** [reply]
I do not not believe that you are correct. actions will take place on all active channels. You need to deactivate any channels with +channel before processing some other channel. I also do not understand your use of L=1,A=2 and B=4 for IM 6.9.7.4. Nothing has changed in regard to channel numbering from that to 6.9.10-23.

**Reply by fmw42** [reply]
I am wrong. These two do produce the same results.

![lena](https://user-images.githubusercontent.com/13458218/119550728-98e1d300-bd4d-11eb-93c7-4ddd537d5532.png)


```
convert lena.png -channel r -negate -channel g -negate +channel x1.png
convert lena.png -channel r -negate +channel -channel g -negate +channel x2.png
```

![x1](https://user-images.githubusercontent.com/13458218/119550750-9e3f1d80-bd4d-11eb-9550-3ab57b3d0721.png)
